kubeadm: add v1beta4 API boilerplate

neolit123 opened the pull request:
/kind feature

#### What this PR does / why we need it:

Create v1beta4 by copying v1beta3.
The API is not added to the scheme, which means it is not allowed as --config input yet.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref https://github.com/kubernetes/kubeadm/issues/2890

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```
**^^^ NOTE  don't include release notes for new features**

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

neolit123 committed:
kubeadm: add v1beta4 API boilerplate

Create v1beta4 by copying v1beta3.

neolit123 commented:
/approve cancel

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -0,0 +1,271 @@
+// This version improves on the v1beta2 format by fixing some minor issues and adding a few new fields.
+//
+// A list of changes since v1beta3:
+//   - TODO
```
all new feature PRs must update this list.

release notes should be NONE everywhere, and we can release all features at once.

neolit123 replied:
i can add the link next to the TODO, yes

neolit123 reviewed (commented):
@SataQiu 
without adding it to the scheme there will be no way for us to test, conversion, defaulting, etc.
i think we should do that and just PR the --config codepath to not accept v1beta4, WDYT?
(unclear how complicated it will be)

pacoxu commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -0,0 +1,271 @@
+// Migration from old kubeadm config versions
+//
+//   - kubeadm v1.15.x and newer can be used to migrate from v1beta1 to v1beta2.
+//   - kubeadm v1.22.x and newer no longer support v1beta1 and older APIs, but can be used to migrate v1beta2 to v1beta4.
+//   - kubeadm v1.27.x and newer no longer support v1beta2 and older APIs.
```
```suggestion
//   - kubeadm v1.22.x and newer can be used to migrate from v1beta2 to v1beta3.
//   - kubeadm v1.27.x and newer no longer support v1beta2 and older APIs, but can be used to migrate v1beta3 to v1beta4.
```

neolit123 replied:
hm, 31 was a search-replace mistake.
but for 32 we cannot say that 1.27 can be used to migrate to v1beta4, because it does not know about it.

pacoxu replied:
Got it.

BTW, can we define a new concept as an experimental version? Then the `kubeadm config validate --experimental ` and `kubeadm config migrate --experimental  ` can use v1beta4 for testing.

neolit123 replied:
for this TODO i will add an action item in the tracking issue so that we don't forget.

neolit123 replied:
> BTW, can we define a new concept as an experimental version? Then the `kubeadm config validate --experimental ` and `kubeadm config migrate --experimental ` can use v1beta4 for testing.

the CLI is GA so everything that we add as a flag must have a good purpose.
i think exposing the config as experimental is OK, but unclear if users will need it...it also depends on how much time v1veta4 will be a WIP. i'm +0 on this one, but if the others think it's a good idea, we can add it.

chendave replied:
> i think exposing the config as experimental is OK, but unclear if users will need it

I'd say warning msg + experimental should be enough.

neolit123 commented:
failing pull-kubernetes-verify

```
panic: duplicate static defaulter defined: &types.Type{Name:types.Name{Package:"k8s.io/kubernetes/cmd/kubeadm/app/apis/bootstraptoken/v1", Name:"BootstrapToken", Path:""}, Kind:"Struct", CommentLines:[]string(nil), SecondClosestCommentLines:[]string(nil), Members:[]types.Member{types.Member{Name:"Token", Embedded:false, CommentLines:[]string{"Token is used for establishing bidirectional trust between nodes and control-planes.", "Used for joining nodes in the cluster."}, Tags:"json:\"token\" datapolicy:\"token\"", Type:(*types.Type)(0xc0251e6540)}, types.Member{Name:"Description", Embedded:false, CommentLines:[]string{"Description sets a human-friendly message why this token exists and what it's used", "for, so other administrators can know its purpose.", "+optional"}, Tags:"json:\"description,omitempty\"", Type:(*types.Type)(0x9b65c0)}, types.Member{Name:"TTL", Embedded:false, CommentLines:[]string{"TTL defines the time to live for this token. Defaults to 24h.", "Expires and TTL are mutually exclusive.", "+optional"}, Tags:"json:\"ttl,omitempty\"", Type:(*types.Type)(0xc024af3a40)}, types.Member{Name:"Expires", Embedded:false, CommentLines:[]string{"Expires specifies the timestamp when this token expires. Defaults to being set", "dynamically at runtime based on the TTL. Expires and TTL are mutually exclusive.", "+optional"}, Tags:"json:\"expires,omitempty\"", Type:(*types.Type)(0xc0246c8900)}, types.Member{Name:"Usages", Embedded:false, CommentLines:[]string{"Usages describes the ways in which this token can be used. Can by default be used", "for establishing bidirectional trust, but that can be changed here.", "+optional"}, Tags:"json:\"usages,omitempty\"", Type:(*types.Type)(0xc0244ad740)}, types.Member{Name:"Groups", Embedded:false, CommentLines:[]string{"Groups specifies the extra groups that this token will authenticate as when/if", "used for authentication", "+optional"}, Tags:"json:\"groups,omitempty\"", Type:(*types.Type)(0xc0244ad740)}}, Elem:(*types.Type)(nil), Key:(*types.Type)(nil), Underlying:(*types.Type)(nil), Methods:map[string]*types.Type(nil), Signature:(*types.Signature)(nil), ConstValue:(*string)(nil), Len:0}

goroutine 1 [running]:
k8s.io/gengo/examples/defaulter-gen/generators.getManualDefaultingFunctions(0xc026fa5100, 0xc02520a6c0, 0xc00015a280?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/gengo/examples/defaulter-gen/generators/defaulter.go:186 +0xd9b
k8s.io/gengo/examples/defaulter-gen/generators.Packages(0xc026fa5100, 0xc0001922c0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/gengo/examples/defaulter-gen/generators/defaulter.go:275 +0x106e
k8s.io/gengo/args.(*GeneratorArgs).Execute(0xc0001922c0, 0xc00012c010?, {0x74ec40, 0x6}, 0x782038)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/gengo/args/args.go:212 +0x1a2
main.main()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/defaulter-gen/main.go:69 +0x22e
!!! [0620 09:00:53] Call tree:
!!! [0620 09:00:53]  1: hack/update-codegen.sh:919 codegen::defaults(...)
+++ exit code: 1
+++ error: 1
[0;31mFAILED[0m   verify-codegen.sh	185s
```

i cannot run ./hack/update-codegen.sh locally for some reason.

this is odd
> panic: duplicate static defaulter defined:

SataQiu commented:
> i wonder if we have to rework something around the BootstrapToken to make this work.

Previously, `v1beta2` uses the internal bt `func SetDefaults_BootstrapToken(bt *BootstrapToken)`, but `v1beta3` uses the external package `func SetDefaults_BootstrapToken(bt *bootstraptokenv1.BootstrapToken)`.

I think we should move `SetDefaults_BootstrapToken` into `bootstraptoken/v1` package, WDYT?

neolit123 commented:
i was able to reproduce the problem locally, so i can debug it more later

> I think we should move SetDefaults_BootstrapToken into bootstraptoken/v1 package, WDYT?

that might be it. problem with that is the defaulter of the bootstrap token struct ideally must not import the kubeadm constants package for defaulting. if that is the problem i will see what can be done. perhaps just inline the defaulter.

```

// SetDefaults_BootstrapToken sets the defaults for an individual Bootstrap Token
func SetDefaults_BootstrapToken(bt *bootstraptokenv1.BootstrapToken) {
	if bt.TTL == nil {
		bt.TTL = &metav1.Duration{
			Duration: constants.DefaultTokenDuration,
		}
	}
	if len(bt.Usages) == 0 {
		bt.Usages = constants.DefaultTokenUsages
	}

	if len(bt.Groups) == 0 {
		bt.Groups = constants.DefaultTokenGroups
	}
}
```

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -0,0 +1,273 @@
+//
+//   - kubeadm v1.15.x and newer can be used to migrate from v1beta1 to v1beta2.
+//   - kubeadm v1.22.x and newer no longer support v1beta1 and older APIs, but can be used to migrate v1beta2 to v1beta3.
+//   - kubeadm v1.27.x and newer no longer support v1beta2 and older APIs.
```
we will keep v1beta3 for a while (1 year) we can mark it as deprecated once v1beta4 is out.

neolit123 committed:
kubeadm: move token defaulting and related constants to token package

Move the defaulting of the BootstrapToken type inside the
bootstraptoken/v1 package. This prevents an error where
codegen complains that a defaulter for the type exists in both
the kubeadm v1beta3 and v1beta4.

Adapt kubeadm code to use the defaulter function and constants
that were moved to bootstraptoken/v1.

NOTE: technically this is a breaking change for direct users of
v1beta3/SetDefaults_BootstrapToken().

neolit123 commented:
@SataQiu i've moved the bootstraptoken defaulting in the bootstraptoken/v1 package. this is done in a separate commit here.

neolit123 commented:
cspz should not be affected so this is a flake.
/retest

neolit123 commented:
pinged slack about the capz failures:
https://kubernetes.slack.com/archives/CN0K3TE2C/p1687353098949899

not required so we can do:
```
/skip pull-kubernetes-e2e-capz-windows
```

neolit123 commented:
/skip pull-kubernetes-e2e-capz-windows

neolit123 commented:
/skip

chendave commented:
the change LGTM, but not sure if you want to register v1beta4 schema as you comments above,

> without adding it to the scheme there will be no way for us to test, conversion, defaulting, etc.
i think we should do that and just PR the --config codepath to not accept v1beta4, WDYT?
(unclear how complicated it will be)

Maybe we can just leave that to a follow up review?

neolit123 commented:
> the change LGTM, but not sure if you want to register v1beta4 schema as you comments above,
> 
> > without adding it to the scheme there will be no way for us to test, conversion, defaulting, etc.
> > i think we should do that and just PR the --config codepath to not accept v1beta4, WDYT?
> > (unclear how complicated it will be)
> 
> Maybe we can just leave that to a follow up review?

+1 for adding it in a separate pr. i tested it locally.

pacoxu commented:
/approve
/lgtm
/hold
feel free to unhold

chendave commented:
/lgtm

SataQiu commented on cmd/kubeadm/app/apis/kubeadm/v1beta3/defaults.go:
```diff
@@ -181,23 +181,7 @@ func SetDefaults_BootstrapTokens(obj *InitConfiguration) {
-}
-
-// SetDefaults_BootstrapToken sets the defaults for an individual Bootstrap Token
-func SetDefaults_BootstrapToken(bt *bootstraptokenv1.BootstrapToken) {
```
Although we removed `SetDefaults_BootstrapToken` without a deprecate phase, given that it's just a package replace, I think this change is OK.

/lgtm

neolit123 replied:
indeed, there is no other way without blocking the work on the api

neolit123 commented:
will send the next pr that adds the api to the scheme today.

/hold cancel
